neuralop/utils.py

```diff
@@ -1,42 +1,80 @@
+from typing import Union, Literal, Optional, List, Tuple, Generator
 import torch
+from torch import Tensor
 import wandb
 
+
 # normalization, pointwise gaussian
 class UnitGaussianNormalizer:
-    def __init__(self, x, eps=0.00001, reduce_dim=[0], verbose=True):
+    def __init__(
+        self,
+        x: Union[Tensor, Generator],
+        eps: float = 1e-6,
+        reduce_dim: List[int] = [0],
+        verbose: bool = True,
+    ):
         super().__init__()
-        n_samples, *shape = x.shape
-        self.sample_shape = shape
         self.verbose = verbose
         self.reduce_dim = reduce_dim
 
         # x could be in shape of ntrain*n or ntrain*T*n or ntrain*n*T
-        self.mean = torch.mean(x, reduce_dim, keepdim=True).squeeze(0)
-        self.std = torch.std(x, reduce_dim, keepdim=True).squeeze(0)
+        count, mean, std = self.compute_statistics(x, reduce_dim)
         self.eps = eps
-        
+        self.mean = mean
+        self.std = std
+        self.count = count
+
         if verbose:
-            print(f'UnitGaussianNormalizer init on {n_samples}, reducing over {reduce_dim}, samples of shape {shape}.')
-            print(f'   Mean and std of shape {self.mean.shape}, eps={eps}')
+            print(
+                f"UnitGaussianNormalizer init on {count}, reducing over {reduce_dim}."
+            )
+            print(f"   Mean and std of shape {self.mean.shape}, eps={eps}")
+
+    def compute_statistics(
+        self, x: Union[Tensor, Generator], reduce_dim: List[int] = [0]
+    ) -> Tuple[int, Tensor, Tensor]:
+        """
+        Compute mean and standard deviaion
+        """
+        if isinstance(x, Tensor):
+            mean = torch.mean(x, reduce_dim, keepdim=True).squeeze(0)
+            std = torch.std(x, reduce_dim, keepdim=True).squeeze(0)
+            count = len(x)
+        elif isinstance(x, Generator):
+            first_item = next(x)
```

Comment:
As a minor point I would just have the for loop:
```python
for index, element in enumerate(x):
    if index:
        total_sum += item
        total_sum_square += item**2
    else:
        total_sum = element
        total_sum_square = element**2
mean = total_sum / (index + 1)
variance = ( total_sum_square - (total_sum**2)/(index + 1) )/index
```

Reply:
I think we can also make it a standalone function e.g.
incremental_mean_variance(data_iterator)

Reply:
Lastly maybe we should support batch - this way we can directly call the normalizer on a (lazy) data loader?

Reply:
Oh also shouldn't we also use the reduce dim when computing mean / variance?

Reply:
I will add the support for the reduction. The test is already done on a batch.